torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -441,19 +443,98 @@ void ONNXSetDynamicInputShape(
   }
 }
 
+bool HasSequenceTypeOutput(Node* node) {
+  if (node->kind() == ::c10::onnx::SplitToSequence ||
+      node->kind() == ::c10::onnx::SequenceInsert ||
+      node->kind() == ::c10::onnx::SequenceEmpty ||
+      node->kind() == ::c10::onnx::SequenceErase ||
+      node->kind() == ::c10::onnx::SequenceConstruct)
+    return true;
+  return false;
+}
+
+void ONNXUpdateTypeFromTensor(
+    Value* graph_output,
+    at::Tensor output,
+    bool onnx_shape_inference) {
+  if (onnx_shape_inference) {
+    graph_output->setType(
+        MergeInferredType(TensorType::create(output), graph_output->type()));
+  } else {
+    graph_output->inferTypeFrom(output);
+  }
+}
+
 void ONNXAssignOutputShape(
     std::shared_ptr<Graph>& graph,
     at::ArrayRef<at::Tensor> outputs,
+    const python::IODescriptor& desc,
     bool onnx_shape_inference) {
-  TORCH_INTERNAL_ASSERT(graph->outputs().size() == outputs.size());
-  for (size_t i = 0; i < outputs.size(); ++i) {
-    if (onnx_shape_inference) {
-      graph->outputs()[i]->setType(MergeInferredType(
-          TensorType::create(outputs[i]), graph->outputs()[i]->type()));
-    } else {
-      graph->outputs()[i]->inferTypeFrom(outputs[i]);
+  size_t outputs_index = 0;
+
+  auto py_obj = unflatten(outputs, desc);
+
+  TORCH_INTERNAL_ASSERT(PyTuple_Check(py_obj));
+
+  for (size_t i = 0; i < PyTuple_GET_SIZE(py_obj); ++i) {
+    PyObject* elem = PyTuple_GET_ITEM(py_obj, i);
+
+    if (PyList_Check(elem)) {
+      size_t list_len = PyList_GET_SIZE(elem);
+      if (HasSequenceTypeOutput(graph->outputs()[i]->node())) {
+        if (list_len > 0) {
+          auto& var =
+              reinterpret_cast<THPVariable*>(PyList_GET_ITEM(elem, 0))->cdata;
+          for (size_t j = 1; j < list_len; ++j) {
+            PyObject* list_elem = PyList_GET_ITEM(elem, j);
+            auto& new_var = reinterpret_cast<THPVariable*>(list_elem)->cdata;
+            TORCH_INTERNAL_ASSERT(THPVariable_Check(list_elem));
+            TORCH_INTERNAL_ASSERT(var.scalar_type() == new_var.scalar_type());
+          }
+          outputs_index += list_len;
+          graph->outputs()[i]->setType(ListType::create(
+              TensorType::create(var.scalar_type(), at::kCPU, {}, {})));
+          ONNXUpdateTypeFromTensor(
```

Comment:
Currently `MergeInferredType` does not handle case merging `List[Tensor]` with `Tensor`. In ONNX, different tensor shape inside one Sequence is supported. And we try best effort to find the compatible representation. Perhaps do something similar with the list of `vars` and find a compatible tensor shape?

Reply:
We're inferring the type here. Looks like the output shape for sequences is already handled in export.cpp.